cli/src/commands/git/push.rs

```diff
@@ -295,17 +300,56 @@ pub fn cmd_git_push(
         return Ok(());
     }
 
-    validate_commits_ready_to_push(ui, &bookmark_updates, &remote, &tx, args)?;
+    let sign_behavior = if tx.settings().get_bool("git.sign-on-push")? {
+        Some(SignBehavior::Own)
+    } else {
+        None
+    };
+    let commits_to_sign =
+        validate_commits_ready_to_push(ui, &bookmark_updates, &remote, &tx, args, &sign_behavior)?;
+    let num_updated_signatures = commits_to_sign.len();
+
+    if args.dry_run {
+        if let Some(mut formatter) = ui.status_formatter() {
+            if num_updated_signatures > 0 {
+                writeln!(
+                    formatter,
+                    "Dry-run requested, skipped updating signatures of {num_updated_signatures} \
+                     commits"
+                )?;
+            }
+            writeln!(formatter, "Changes to push to {remote}:")?;
+            print_commits_ready_to_push(formatter.as_mut(), tx.repo(), &bookmark_updates)?;
+            writeln!(formatter, "Dry-run requested, not pushing.")?;
+            return Ok(());
+        }
+    }
+
+    let num_rebased_descendants: usize;
+    if let Some(sign_behavior) = sign_behavior {
+        (num_rebased_descendants, bookmark_updates) =
+            sign_commits_before_push(&mut tx, commits_to_sign, sign_behavior, bookmark_updates)?;
+    } else {
+        num_rebased_descendants = 0;
+    }
+
     if let Some(mut formatter) = ui.status_formatter() {
+        if num_updated_signatures > 0 {
+            writeln!(
+                formatter,
+                "Updated signatures of {num_updated_signatures} commits"
```

Comment:
Do we need to be thorough about describing this? E.g. listing the number of newly signed commits, number of signatures dropped, for both the commits being pushed and the rebased descendants.

Reply:
I'm not sure. This seems fine to me. We can adjust later if people think it's not sufficient.